### DISCUSSION TITLE
[Spring]리포지토리 내 함수 리턴 타입 질문

### DISCUSSION THREAD

**Discussion by DoohyunHwang97** [wafflestudio/seminar-2023#160] [Q&A] [answered]
리포지토리 내 조회하는 함수가 있을 때
리턴 타입은 항상 엔티티 혹은 엔티티 컬렉션이 되어야 하는 것이 맞을까요?
SongRepository의 경우 리턴타입으로 List<Song>을 쓸 수 있는지 궁금합니다.
`fun findByTitleContaining(title: String): List<SongEntity>`

**Comment by davin111** [comment, marked as answer]
@DoohyunHwang97 안녕하세요, Spring Data 의 repository 에 대해서 일반적인 질문이라면(사실 repository 라는 용어 자체는 설계상 계층적인 개념일 뿐이고 어떻게 구현할지는 맘대로니까요), 다음 문서 내용을 참고하실 수 있습니다. https://docs.spring.io/spring-data/commons/docs/current/reference/html/#appendix.query.return.types

말씀하신대로, 일반적으로 entity (table 과 매핑한 class) 내지는 entity 의 컬렉션을 제일 많이 쓰시게 될 거구요. 그러나 꽤 다양한 return type 들이 추가로 가능하답니다. 물론 문서에도 나와있듯, DB 특성 등에 따라 지원되는 return type 이 조금씩은 달라질 수 있습니다.

`SongRepository `라고 칭하신 것은, 아마 `JpaRepository` interface 를 상속하는 interface 겠죠? 그렇다면 맞습니다. Spring Data 의 repository interface 를 상속해 method 이름으로 쿼리를 생성하는 경우, return type 을 `List<T>`로 못 받는 경우는 거의 없다고 보시면 됩니다.

**Reply by DoohyunHwang97** [reply]
@davin111 답변 감사합니다. 이해가 되었습니다 :)

**Reply by DoohyunHwang97** [reply]
@davin111 추가적인 질문 드립니다!
1. 레포지토리 함수 리턴타입으로 T 타입을 쓰는 것과 Optional<T> 중에 어떤 것이 선호되고 각각 어떤 장점이 있는지 궁금합니다 :)
2. 널 체크를 할 때 T != null 을 사용하는 것과 Optional의 isPresent()로 체크하는 것 중에 어떤 것이 좋을까요?

**Reply by davin111** [reply]
@DoohyunHwang97 cc. @PFCJeong 

T 타입을 쓰는 것이 generic 에 대한 질문이 아니라, 그냥 entity 자체를 return 타입으로 받는 것으로 표현하셨다고 이해하겠습니다.

만약 Java 로 개발을 하는 상황이라면 Optional 을 선호한다고 말할 수 있겠습니다. Java 8 버전부터 `Optional` 이 도입되어, null 체크를 상대적으로 안전하게 할 수 있게 되었습니다. 그 전에는 예를 들어 `User`라는 type 을 다룬다 하더라도, 이게 사실은 null 일 수 있어서 확인하지 않고 그대로 해당 변수의 멤버나 메서드에 접근하면 NullPointerException 이 터지기 쉬웠거든요. 그런데 Optional 을 통해서 실수를 줄이고, 이걸 체크하고 써야한다는 강제를 주는 거죠.

그러나 우리는 Kotlin 을 쓰고 있고, Kotlin 은 언어적으로 null 을 다루는 더 좋은 방법들을 갖고 있습니다. https://kotlinlang.org/docs/null-safety.html 대표적인 것이 바로 `?`인데요. Kotiln 에서 `User`라는 type 은 Java 와 달리, null 이 아니라는 전제를 갖고 있습니다. 만약에 null 일 수도 있다면 `User?`여야 하는 거죠. 즉, Java 에 뒤늦게 도입된 Optional 을 초기부터 내장하고 있었다고 볼 수 있습니다. 그렇기에 Kotlin 프로그래밍을 하실 때 Optional 은 더이상 필요가 없습니다.

Spring Data 의 repository 에서도 Kotlin 지원의 일환으로 return type `T?`를 가능하게 해주고 있습니다. https://docs.spring.io/spring-data/commons/docs/current/reference/html/#repositories.nullability.kotlin 그러니까 결과물이 없을 수 있는 메서드의 경우에는 Kotlin 에서 `T?` 를 쓰는 게 적절합니다.

* * *

위의 답변을 통해 1.과 2.가 한 번에 답변된 셈이라 할 수 있을 거 같아요. Kotlin 에서 null 체크는 T != null 도 있겠지만, `T?.let { }` 같은 식의 scope 함수를 사용하는 등 다양하고 유창한 여러 방식이 가능합니다.

**Reply by DoohyunHwang97** [reply]
@davin111 올려주신 링크를 통해 T?에 대해서 봤는데 너무 편하고 신기했습니다. Optional이 제공하는 orElse나 orElseThrow 같은 기능들 때문에 Optional이 더 유용하지 않나 이런 생각이 있었는데 Elvis Operator 보고 바로 생각이 바뀌었습니다 ㅋㅋ 덕분에 또 하나 알아갑니다. 감사합니당
